{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import muon as mu\n",
    "\n",
    "\n",
    "# Add the Functions folder to your path\n",
    "sys.path.append(os.path.abspath(\"/Users/alexandra/Desktop/EG-CLR/Functions\"))\n",
    "\n",
    "# Now import your functions\n",
    "from QC_functions import *\n",
    "from CLR_functions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/CLR/lib/python3.11/site-packages/mudata/_core/mudata.py:1531: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/opt/anaconda3/envs/CLR/lib/python3.11/site-packages/mudata/_core/mudata.py:1429: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "# Load multimodal data\n",
    "mdata = mu.read(\"/Volumes/G-DRIVE mobile USB-C/Single-cell_data/K562/10x/ISSAACC-seq_generated/hg19_10xCloud_aligned_data/processed_data/multiome_data.h5mu\")\n",
    "\n",
    "adata_gem = mdata['rna']\n",
    "adata_atac = mdata['atac']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain orginal read counts\n",
    "adata_org = sc.read_10x_mtx('/Volumes/G-DRIVE mobile USB-C/Single-cell_data/K562/10x/ISSAACC-seq_generated/hg19_10xCloud_aligned_data/filtered_feature_bc_matrix',\n",
    "                            gex_only = False)\n",
    "\n",
    "gex_rows = list(map(lambda x: x == 'Gene Expression', adata_org.var['feature_types']))\n",
    "atac_rows = list(map(lambda x: x == 'Peaks', adata_org.var['feature_types']))\n",
    "\n",
    "adata_gem_org = adata_org[:, gex_rows].copy()\n",
    "adata_atac_org = adata_org[:, atac_rows].copy()\n",
    "\n",
    "adata_gem_org.var_names_make_unique()\n",
    "adata_atac_org.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace read counts with orgaianl read counts\n",
    "gem_index = adata_gem.var_names\n",
    "atac_index = adata_atac.var.index\n",
    "\n",
    "cell_index = adata_gem.obs.index\n",
    "\n",
    "adata_gem_new = adata_gem_org[:, adata_gem_org.var_names.isin(gem_index)].copy()\n",
    "adata_atac_new = adata_atac_org[:, adata_atac_org.var_names.isin(atac_index)].copy()\n",
    "\n",
    "adata_gem_new = adata_gem_new[adata_gem_new.obs_names.isin(cell_index)].copy()\n",
    "adata_atac_new = adata_atac_new[adata_atac_new.obs_names.isin(cell_index)].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 10359 × 503\n",
       "    var: 'gene_ids', 'feature_types', 'Chromosome', 'Start', 'End'\n",
       "    uns: 'log1p'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# normalizing scRNA-seq by chromasomex \n",
    "adata_gem_new = assign_chr('/Volumes/G-DRIVE mobile USB-C/Single-cell_data/K562/10x/ISSAACC-seq_generated/hg19_10xCloud_aligned_data/filtered_feature_bc_matrix/features.tsv.gz', adata_gem_new)\n",
    "adata_gem_chrx = subset_adata_by_chromosome(adata_gem_new,-1)\n",
    "sc.pp.normalize_total(adata_gem_chrx, target_sum=1e4)\n",
    "sc.pp.log1p(adata_gem_chrx)\n",
    "adata_gem_chrx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 10359 × 3424\n",
       "    var: 'gene_ids', 'feature_types', 'gene', 'distance', 'peak_type', 'Chromosome', 'Start', 'End'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define promoter, enhancer and gene body regions\n",
    "adata_atac_new = assign_loc('/Volumes/G-DRIVE mobile USB-C/Single-cell_data/K562/10x/ISSAACC-seq_generated/hg19_10xCloud_aligned_data/filtered_feature_bc_matrix/features.tsv.gz',\n",
    "    \"/Volumes/G-DRIVE mobile USB-C/Single-cell_data/K562/10x/ISSAACC-seq_generated/hg19_10xCloud_aligned_data/atac_peak_annotation.tsv\", adata_atac_new)\n",
    "adata_atac_new = assign_chr('/Volumes/G-DRIVE mobile USB-C/Single-cell_data/K562/10x/ISSAACC-seq_generated/hg19_10xCloud_aligned_data/filtered_feature_bc_matrix/features.tsv.gz', adata_atac_new)\n",
    "adata_atac_chrx = subset_adata_by_chromosome(adata_atac_new,-1)\n",
    "adata_atac_chrx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# binarize scATAC-seq \n",
    "adata_atac_chrx.X = (adata_atac_chrx.X > 0).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute CLR matrix for GATA1 -> ChrX\n",
    "MI_Matrix_chrx = MI_Matrix_MIinfoClassif(adata_gem_chrx, adata_atac_chrx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLR_matrix_chrx = CLR_Matrix(MI_Matrix_chrx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write_matrixes(\"ChrX_allgene_allpeaks\",GATA1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3424 × 503\n",
       "    obs: 'gene_ids', 'feature_types', 'gene', 'distance', 'peak_type', 'Chromosome', 'Start', 'End'\n",
       "    var: 'gene_ids', 'feature_types', 'Chromosome', 'Start', 'End', 'gene'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_CLR_matrix_chrx = ad.AnnData(X=CLR_matrix_chrx.values)\n",
    "adata_CLR_matrix_chrx.obs = adata_atac_chrx.var.copy()\n",
    "adata_CLR_matrix_chrx.var = adata_gem_chrx.var.copy()\n",
    "adata_CLR_matrix_chrx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chrX:62333-63291            0.000000\n",
       "chrX:63937-64751            1.680467\n",
       "chrX:68722-69507            0.000000\n",
       "chrX:69671-70530            0.083656\n",
       "chrX:162739-163466          0.000000\n",
       "                              ...   \n",
       "chrX:155249188-155250096    1.749103\n",
       "chrX:155253186-155254082    0.000000\n",
       "chrX:155254614-155255426    1.475307\n",
       "chrX:155258082-155258986    0.000000\n",
       "chrX:155259726-155260695    1.619834\n",
       "Name: GATA1, Length: 3424, dtype: float64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLR_matrix_chrx = pd.DataFrame(CLR_matrix_chrx.values, index=adata_atac_chrx.var['gene'], columns=adata_gem_chrx.var['geng_ids'])\n",
    "GATA1 = CLR_matrix_chrx[\"GATA1\"].copy()\n",
    "GATA1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a gene name and a CLR matrix, and returns a DataFrame containing the CLR values for the specified genes.\n",
    "def EG_pair_by_name(gene, CLR_Matrix):\n",
    "    \n",
    "    if gene not in CLR_Matrix.columns:\n",
    "        print(f\"Warning: {gene} not found in CLR matrix.\")\n",
    "    else:\n",
    "        EG_pair = CLR_Matrix[gene].copy()\n",
    "        EG_pair = EG_pair[EG_pair > 0]\n",
    "        #GATA1_EG_pair = GATA1_EG_pair.reset_index()\n",
    "        \n",
    "        \n",
    "        #GATA1_EG_pair.columns = ['Peak', 'CLR_value']\n",
    "        GATA1_EG_pair = EG_pair.index.str.extract(r'^(chr\\w+):(\\d+)-(\\d+)$')\n",
    "        GATA1_EG_pair.columns = ['chr', 'start', 'end']\n",
    "\n",
    "    \n",
    "        return GATA1_EG_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>63937</td>\n",
       "      <td>64751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>69671</td>\n",
       "      <td>70530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167850</td>\n",
       "      <td>168779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>176215</td>\n",
       "      <td>177159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>192744</td>\n",
       "      <td>193400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155232071</td>\n",
       "      <td>155232857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1954</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155240875</td>\n",
       "      <td>155241468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1955</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155249188</td>\n",
       "      <td>155250096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1956</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155254614</td>\n",
       "      <td>155255426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1957</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155259726</td>\n",
       "      <td>155260695</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1958 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr      start        end\n",
       "0     chrX      63937      64751\n",
       "1     chrX      69671      70530\n",
       "2     chrX     167850     168779\n",
       "3     chrX     176215     177159\n",
       "4     chrX     192744     193400\n",
       "...    ...        ...        ...\n",
       "1953  chrX  155232071  155232857\n",
       "1954  chrX  155240875  155241468\n",
       "1955  chrX  155249188  155250096\n",
       "1956  chrX  155254614  155255426\n",
       "1957  chrX  155259726  155260695\n",
       "\n",
       "[1958 rows x 3 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GATA1_EG_pair = EG_pair_by_name(\"GATA1\", CLR_matrix_chrx)\n",
    "GATA1_EG_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/CLR/lib/python3.11/site-packages/openpyxl/worksheet/header_footer.py:48: UserWarning: Cannot parse header or footer so it will be ignored\n",
      "  warn(\"\"\"Cannot parse header or footer so it will be ignored\"\"\")\n",
      "/opt/anaconda3/envs/CLR/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>class</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48641372</td>\n",
       "      <td>48641493</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48659028</td>\n",
       "      <td>48659074</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48660375</td>\n",
       "      <td>48660526</td>\n",
       "      <td>promoter</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1848</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46655945</td>\n",
       "      <td>46656445</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1849</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46696085</td>\n",
       "      <td>46696605</td>\n",
       "      <td>promoter</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2117</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49962985</td>\n",
       "      <td>49963545</td>\n",
       "      <td>genic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49969065</td>\n",
       "      <td>49969605</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2119</th>\n",
       "      <td>chrX</td>\n",
       "      <td>50191545</td>\n",
       "      <td>50192045</td>\n",
       "      <td>genic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2120</th>\n",
       "      <td>chrX</td>\n",
       "      <td>50322885</td>\n",
       "      <td>50323405</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2121</th>\n",
       "      <td>chrX</td>\n",
       "      <td>50434665</td>\n",
       "      <td>50435165</td>\n",
       "      <td>genic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>277 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start       end       class  Significant\n",
       "196   chrX  48641372  48641493  intergenic         True\n",
       "197   chrX  48659028  48659074  intergenic         True\n",
       "198   chrX  48660375  48660526    promoter         True\n",
       "1848  chrX  46655945  46656445  intergenic        False\n",
       "1849  chrX  46696085  46696605    promoter        False\n",
       "...    ...       ...       ...         ...          ...\n",
       "2117  chrX  49962985  49963545       genic        False\n",
       "2118  chrX  49969065  49969605  intergenic        False\n",
       "2119  chrX  50191545  50192045       genic        False\n",
       "2120  chrX  50322885  50323405  intergenic        False\n",
       "2121  chrX  50434665  50435165       genic        False\n",
       "\n",
       "[277 rows x 5 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CRISPRi_data = pd.read_excel('/Users/alexandra/Desktop/Data/CRISPRiFlowFISH/41588_2019_538_MOESM3_ESM.xlsx',\n",
    "                             sheet_name='Supplementary Table 6a', \n",
    "                             skiprows=0,\n",
    "                             header=1)\n",
    "CRISPRi_GATA1 = CRISPRi_data[CRISPRi_data['Gene'] == 'GATA1'].copy()\n",
    "CRISPRi_GATA1 = CRISPRi_GATA1[['chr', 'start', 'end', 'class', 'Significant']].copy()\n",
    "CRISPRi_GATA1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "CRISPRi_GATA1_TP = CRISPRi_GATA1[CRISPRi_GATA1[\"Significant\"] == True]\n",
    "CRISPRi_GATA1_TN = CRISPRi_GATA1[CRISPRi_GATA1[\"Significant\"] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48641372</td>\n",
       "      <td>48641493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48659028</td>\n",
       "      <td>48659074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48660375</td>\n",
       "      <td>48660526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1993</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48619903</td>\n",
       "      <td>48620645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48643025</td>\n",
       "      <td>48643685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2003</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48700065</td>\n",
       "      <td>48700565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2065</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49040725</td>\n",
       "      <td>49041225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49100665</td>\n",
       "      <td>49101185</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start       end\n",
       "196   chrX  48641372  48641493\n",
       "197   chrX  48659028  48659074\n",
       "198   chrX  48660375  48660526\n",
       "1993  chrX  48619903  48620645\n",
       "1999  chrX  48643025  48643685\n",
       "2003  chrX  48700065  48700565\n",
       "2065  chrX  49040725  49041225\n",
       "2081  chrX  49100665  49101185"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CRISPRi_GATA1_TP = CRISPRi_GATA1_TP[['chr', 'start', 'end']].copy()\n",
    "CRISPRi_GATA1_TP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>class</th>\n",
       "      <th>Significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1848</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46655945</td>\n",
       "      <td>46656445</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1849</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46696085</td>\n",
       "      <td>46696605</td>\n",
       "      <td>promoter</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46696685</td>\n",
       "      <td>46697185</td>\n",
       "      <td>promoter</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1851</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46697525</td>\n",
       "      <td>46698025</td>\n",
       "      <td>genic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1852</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46771460</td>\n",
       "      <td>46772025</td>\n",
       "      <td>promoter</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2117</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49962985</td>\n",
       "      <td>49963545</td>\n",
       "      <td>genic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49969065</td>\n",
       "      <td>49969605</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2119</th>\n",
       "      <td>chrX</td>\n",
       "      <td>50191545</td>\n",
       "      <td>50192045</td>\n",
       "      <td>genic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2120</th>\n",
       "      <td>chrX</td>\n",
       "      <td>50322885</td>\n",
       "      <td>50323405</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2121</th>\n",
       "      <td>chrX</td>\n",
       "      <td>50434665</td>\n",
       "      <td>50435165</td>\n",
       "      <td>genic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>269 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start       end       class  Significant\n",
       "1848  chrX  46655945  46656445  intergenic        False\n",
       "1849  chrX  46696085  46696605    promoter        False\n",
       "1850  chrX  46696685  46697185    promoter        False\n",
       "1851  chrX  46697525  46698025       genic        False\n",
       "1852  chrX  46771460  46772025    promoter        False\n",
       "...    ...       ...       ...         ...          ...\n",
       "2117  chrX  49962985  49963545       genic        False\n",
       "2118  chrX  49969065  49969605  intergenic        False\n",
       "2119  chrX  50191545  50192045       genic        False\n",
       "2120  chrX  50322885  50323405  intergenic        False\n",
       "2121  chrX  50434665  50435165       genic        False\n",
       "\n",
       "[269 rows x 5 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CRISPRi_GATA1_TN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRISPR for GATA1 starts at  46655945  end at  50435165\n"
     ]
    }
   ],
   "source": [
    "CRISPRi_GATA1_start =  CRISPRi_GATA1['start'].min()\n",
    "CRISPRi_GATA1_end = CRISPRi_GATA1['end'].max() \n",
    "\n",
    "print(\"CRISPR for GATA1 starts at \",CRISPRi_GATA1_start, \" end at \", CRISPRi_GATA1_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46762892</td>\n",
       "      <td>46763815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46771506</td>\n",
       "      <td>46772278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46772417</td>\n",
       "      <td>46772801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46828749</td>\n",
       "      <td>46829653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46907130</td>\n",
       "      <td>46908041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49686778</td>\n",
       "      <td>49687624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49901127</td>\n",
       "      <td>49901946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49933733</td>\n",
       "      <td>49934640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>751</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49944910</td>\n",
       "      <td>49945812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49962783</td>\n",
       "      <td>49963685</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr     start       end\n",
       "597  chrX  46762892  46763815\n",
       "598  chrX  46771506  46772278\n",
       "599  chrX  46772417  46772801\n",
       "600  chrX  46828749  46829653\n",
       "601  chrX  46907130  46908041\n",
       "..    ...       ...       ...\n",
       "748  chrX  49686778  49687624\n",
       "749  chrX  49901127  49901946\n",
       "750  chrX  49933733  49934640\n",
       "751  chrX  49944910  49945812\n",
       "752  chrX  49962783  49963685\n",
       "\n",
       "[156 rows x 3 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for overlapping peaks\n",
    "GATA1_EG_pair_limited = GATA1_EG_pair[\n",
    "    (GATA1_EG_pair[\"end\"].values.astype(int) < CRISPRi_GATA1_end ) & \n",
    "    (GATA1_EG_pair[\"start\"].values.astype(int) > CRISPRi_GATA1_start)\n",
    "]\n",
    "GATA1_EG_pair_limited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46762892</td>\n",
       "      <td>46763815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46771506</td>\n",
       "      <td>46772278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46772417</td>\n",
       "      <td>46772801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46828749</td>\n",
       "      <td>46829653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>chrX</td>\n",
       "      <td>46907130</td>\n",
       "      <td>46908041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49686778</td>\n",
       "      <td>49687624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49901127</td>\n",
       "      <td>49901946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49933733</td>\n",
       "      <td>49934640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>751</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49944910</td>\n",
       "      <td>49945812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49962783</td>\n",
       "      <td>49963685</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr     start       end\n",
       "597  chrX  46762892  46763815\n",
       "598  chrX  46771506  46772278\n",
       "599  chrX  46772417  46772801\n",
       "600  chrX  46828749  46829653\n",
       "601  chrX  46907130  46908041\n",
       "..    ...       ...       ...\n",
       "748  chrX  49686778  49687624\n",
       "749  chrX  49901127  49901946\n",
       "750  chrX  49933733  49934640\n",
       "751  chrX  49944910  49945812\n",
       "752  chrX  49962783  49963685\n",
       "\n",
       "[156 rows x 3 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GATA1_EG_pair_limited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48641372</td>\n",
       "      <td>48641493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48659028</td>\n",
       "      <td>48659074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48660375</td>\n",
       "      <td>48660526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1993</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48619903</td>\n",
       "      <td>48620645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48643025</td>\n",
       "      <td>48643685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2003</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48700065</td>\n",
       "      <td>48700565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2065</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49040725</td>\n",
       "      <td>49041225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49100665</td>\n",
       "      <td>49101185</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start       end\n",
       "196   chrX  48641372  48641493\n",
       "197   chrX  48659028  48659074\n",
       "198   chrX  48660375  48660526\n",
       "1993  chrX  48619903  48620645\n",
       "1999  chrX  48643025  48643685\n",
       "2003  chrX  48700065  48700565\n",
       "2065  chrX  49040725  49041225\n",
       "2081  chrX  49100665  49101185"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "CRISPRi_GATA1_TP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t100\t150\n",
      "chr1\t200\t250\n",
      "chr2\t300\t350\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pybedtools\n",
    "\n",
    "# Example DataFrame\n",
    "df = pd.DataFrame({\n",
    "    \"chr\": [\"chr1\", \"chr1\", \"chr2\"],\n",
    "    \"start\": [100, 200, 300],\n",
    "    \"end\": [150, 250, 350]\n",
    "})\n",
    "\n",
    "# Convert to BedTool\n",
    "bed_str = df.to_csv(sep=\"\\t\", header=False, index=False)\n",
    "bed = pybedtools.BedTool(bed_str, from_string=True)\n",
    "\n",
    "# Now you can use .intersect(), .saveas(), etc.\n",
    "print(bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr1\\t100\\t150\\nchr1\\t200\\t250\\nchr2\\t300\\t350\\n'"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48619991</td>\n",
       "      <td>48620883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48640998</td>\n",
       "      <td>48641905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48658644</td>\n",
       "      <td>48659432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48659993</td>\n",
       "      <td>48660815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>49040546</td>\n",
       "      <td>49041486</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr     start       end\n",
       "0  chrX  48619991  48620883\n",
       "1  chrX  48640998  48641905\n",
       "2  chrX  48658644  48659432\n",
       "3  chrX  48659993  48660815\n",
       "4  chrX  49040546  49041486"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load as BedTool\n",
    "GATA1_EG_pair_limited_bed_str = GATA1_EG_pair_limited.to_csv(sep=\"\\t\", header=False, index=False)\n",
    "CRISPRi_GATA1_TP_bed_str = CRISPRi_GATA1_TP.to_csv(sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "GATA1_EG_pair_limited_bed = pybedtools.BedTool(GATA1_EG_pair_limited_bed_str, from_string=True)\n",
    "CRISPRi_GATA1_TP_bed = pybedtools.BedTool(CRISPRi_GATA1_TP_bed_str, from_string=True)\n",
    "\n",
    "# Intersect: return entries in A that overlap with B\n",
    "overlap = GATA1_EG_pair_limited_bed.intersect(CRISPRi_GATA1_TP_bed, u=True)  # 'u=True' gives only entries from A\n",
    "\n",
    "# Convert result to DataFrame\n",
    "overlap_df = overlap.to_dataframe(names=[\"chr\", \"start\", \"end\"])\n",
    "overlap_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# remove peaks that does not exist at the begining"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CLR",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
